### PR TITLE
[SYCL-MLIR] Construct array member in-place in initializer lists

### PR DESCRIPTION
**Pull request by victor-eds**
Construct this members in-place instead of trying to allocate the array, constructing it and then copying.

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 1]
```diff
@@ -367,6 +367,18 @@ mlir::Attribute MLIRScanner::InitializeValueByInitListExpr(mlir::Value ToInit,
       return mlir::DenseElementsAttr();
     }
 
+    if (auto *AIL = dyn_cast<ArrayInitLoopExpr>(Expr)) {
+      // In-house bitcast from `memref<-xarray<N x Ty>>` to `memref<? x Ty>`
+      ValueCategory ArrayPtr(ToInit, /*IsReference=*/true, ElementTy);
+      mlir::Type ET = cast<LLVM::LLVMArrayType>(ElementTy).getElementType();
+      ArrayPtr = ArrayPtr.MemRef2Ptr(Builder, Loc);
+      ArrayPtr.ElementType = ET;
+      ArrayPtr = ArrayPtr.Ptr2MemRef(Builder, Loc);
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:376]: When lowering, this will turn into a `nop` with opaque pointers and a `bitcast` with typed pointers

**Comment by victor-eds** [comment]
@jopperm, wanna take a look at this?

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 2]
```diff
@@ -367,6 +367,18 @@ mlir::Attribute MLIRScanner::InitializeValueByInitListExpr(mlir::Value ToInit,
       return mlir::DenseElementsAttr();
     }
 
+    if (auto *AIL = dyn_cast<ArrayInitLoopExpr>(Expr)) {
+      // In-house bitcast from `memref<-xarray<N x Ty>>` to `memref<? x Ty>`
```
Comment (jopperm) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:371]: Nit: `-` in `memref<..>` (or does that have a special meaning?)

**Review by jopperm — APPROVED** [review]
LGTM!

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 2 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:371]: I meant it could be `?` or a `number`

Comment (jopperm) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:371]: I think it's fine then, bummer that `memref` already uses all the good placeholder chars 🤔

**Review by etiotto — APPROVED** [review]

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 2 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc]: Yes, they left us no good choice...

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 3]
```diff
@@ -623,13 +623,22 @@ ValueCategory MLIRScanner::CommonArrayToPointer(ValueCategory Scalar) {
   }
 
   auto MT = cast<MemRefType>(Scalar.val.getType());
+
+  mlir::Type ET = MT.getElementType();
+  if (auto AT = dyn_cast<LLVM::LLVMArrayType>(ET)) {
+    // In-house bitcast from `memref<-xarray<N x Ty>>` to `memref<? x Ty>`
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:629]: nit:
```suggestion
    // In-house bitcast from `memref<- x array<N x Ty>>` to `memref<? x Ty>`
```

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**